<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>CABANA Workshop: Generating a HiC matrix</title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="PBI - Transcriptomica" href="http://bi.ira.cinvestav.mx:8282/"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      <div class="banner">
        <a href="https://liz-fernandez.github.io/HiC-Langebio/" title="HiC Workshop">
          <img alt="Logo Cinvestav-Royal" src="img/cinvestav_Royal.jpg" />
        </a>
      </div>
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <h1 class="title">Generating a HiC matrix</h1>
          <h2 id="learning-objectives" class="objectives">Learning objectives</h2>
<ul>
<li>Use juicer tools to obtain a multi resolution hic matrix (for juicebox visualization)</li>
<li>Use cooler to obtain a multi resolution hic matrix (for HiCexplorer and HiGlass)</li>
</ul>
<h2 id="hic-.pairs-data-format">HiC .pairs data format</h2>
<p>In this section we will use our filtered hic read pairs in bam format, and convert it to a data format commonly used by downstream HiC analysis tools.</p>
<p>First, we will obtain a .pairs format.</p>
<p>.pairs is a standard format based on plain text proposed by the <a href="https://commonfund.nih.gov/4dnucleome">4DNucleome consortium</a>.</p>
<p>Here is an example:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## pairs format v1.0</span>
<span class="co">#sorted: chr1-chr2-pos1-pos2</span>
<span class="co">#shape: upper triangle</span>
<span class="co">#genome_assembly: hg38</span>
<span class="co">#chromsize: chr1 249250621</span>
<span class="co">#chromsize: chr2 243199373</span>
<span class="co">#chromsize: chr3 198022430</span>

<span class="co">#columns: readID chr1 pos1 chr2 pos2 strand1 strand2</span>
<span class="kw">EAS139</span>:136:FC706VJ:2:2104:23462:197393 chr1 10000 chr1 20000 + +
<span class="kw">EAS139</span>:136:FC706VJ:2:8762:23765:128766 chr1 50000 chr1 70000 + +
<span class="kw">EAS139</span>:136:FC706VJ:2:2342:15343:9863 chr1 60000 chr2 10000 + +
<span class="kw">EAS139</span>:136:FC706VJ:2:1286:25:275154 chr1 30000 chr3 40000 + -</code></pre></div>
<p>HiC data is often quite large, which means that it is not very efficient to work directly with plain text files. Tools for compression and indexing are available for the .pairs format. Some downstream analysis tools accept compressed and indexed files, which makes storage and access easier.</p>
<p>You can read more about the specification and supporting tools here: https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md</p>
<p>We will use the bam2pairs command to obtain a .pairs file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">bam2pairs</span> -l -c chr_file.txt ZmMC_HiC_1_1_2.hicup.bam ZmMC_HiC_1_1_2.hicup</code></pre></div>
<p>Let’s see how the .pairs format looks like:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">head</span> ZmMC_HiC_1_1_2.hicup.bsorted.pairs</code></pre></div>
<h2 id="binning-and-matrix-formats">Binning and matrix formats</h2>
<p>The next step in the workflow is to aggregate the read level pairs into bins. Once the data is binned, other formats are used to store the matrix data.</p>
<p>We will bin and store matrices using two common tools: juicer_tools, which stores the resulting matrix in the .hic format, and cooler, which uses the .cool format.</p>
<p>Both formats are binary and compressed.</p>
<h3 id="hic">.hic</h3>
<p>The .hic is mainly supported by juicer, juicer_tools, and juicebox from the <a href="https://github.com/aidenlab/juicer">Aiden lab</a>. It is a binary format that stores a HiC matrix with multiple bin sizes and corrections in a single file.</p>
<p>Let’s convert our .pairs file to a .hic file.</p>
<p>The juicer_tools pre command bins and corrects the HiC matrix at several resolutions.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">java</span> -Xmx1G -jar juicer_tools_1.13.02.jar pre ZmMC_HiC_1_1_2.hicup.bsorted.pairs ZmMC_HiC_1_1_2.hicup.hic chr_file.txt</code></pre></div>
<p>The resulting .hic file stores the matrix at various resolutions, as well as different corrections. This file can be directly uploaded to juicebox for visualization, and is ready to use with juicer_tools.</p>
<h3 id="cool">.cool</h3>
<p>Another common format is .cool. This format is based on the hdf5 (hierachical data format 5) specification, which is also a binary format. There are multiple open source libraries for accessing and writing hdf5 files available for many programming languages.</p>
<p>.cool files are compatible with HiCExplorer and cooler, common HiC data processing resources. Let’s convert our .pairs file to a .cool file binned at 10Kb resolution.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cooler</span> cload pairs -c1 2 -p1 3 -c2 4 -p2 5 chr_file.txt:10000 ZmMC_HiC_1_1_2.hicup.bsorted.pairs ZmMC_HiC_1_1_2.hicup.10kb.cool</code></pre></div>
<p>Unlike the .hic format, .cool files store a single matrix at a single resolution. However, the .mcool format can store multiresolution matrices. The HiGlass browser supports .mcool matrices.</p>
<p>Let’s obtain a multi resolution matrix with cooler zoomify command.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cooler</span> zoomify ZmMC_HiC_1_1_2.hicup.10kb.cool -r 10000,100000,500000 -o ZmMC_HiC_1_1_2.hicup.mcool</code></pre></div>
<h2 id="matrix-correction">Matrix correction</h2>
<p>Matrix correction is necesary to remove biases like gc content or mappability.</p>
<p>In order to correct a matrix, it is assumed that if no biases were affecting the experiment, each bin should have equal “visibility” of contacts.</p>
<p>This translates to an intuitive solution to matrix correction: transforming the matrix in such a way that the total number of contacts of every row and every column is the same.</p>
<p>Such a procedure is called “matrix balancing”, and many algorithms have been described to achieve this for applications outside HiC data analysis.</p>
<p>For HiC data, the most common ones are called Knight-Ruiz (KR), and Iterative Correction (ICE).</p>
<p>https://liorpachter.wordpress.com/2013/11/17/imakaev_explained/</p>
<p>Before matrix correction, bin level filtering is necesary to remove low count bins.</p>
<p>An approach is the “MAD-max” filter.</p>
<p>To decide filtering values for this filter we can run a diagnostic plot of a histogram of counts per bin, together with the MAD statistic.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicCorrectMatrix</span> diagnostic_plot --matrix ZmMC_HiC_1_1_2.hicup.mcool::/resolutions/10000 -o ZmMC_HiC_1_1_2_diagnostic_10kb.png</code></pre></div>
<p>Let’s explore the diagnostic plot and choose the cutoff values.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">open</span> ZmMC_HiC_2.1.10_sub_1_2_diagnostic_10kb.png</code></pre></div>
<p>After deciding on minimum and maximum values, we can proceed to correct the matrix.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicCorrectMatrix</span> correct --matrix ZmMC_HiC_2.1.10_sub_1_2.hicup.10kb.cool --correctionMethod ICE --outFileName ZmMC_HiC_2.1.10_sub_1_2.hicup.10kb.corrected.cool --filterThreshold -1 5</code></pre></div>
<h2 id="challenge">Challenge:</h2>
<ul>
<li>Obtain and correct matrices at 10kb and 100kb bin sizes for all other replicates and conditions.</li>
</ul>
<h2 id="matrix-qc">Matrix QC</h2>
<p>When working with an experimental design that has multiple conditions and replicates, it is useful to assess how similar the replicates are, and how different the conditions are.</p>
<p>One approach for global comparison is to plot the linear distance versus the contact frequency. The interaction frequency should decay with distance. However, the shape of the decay curve is often characteristic of a single experimental condition. Replicates should have similar decay curves, while conditions should have different ones.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicPlotDistVsCounts</span> --matrices ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool -o plot_vs_counts.png </code></pre></div>
<h2 id="sum-matrices">Sum matrices</h2>
<p>A common practice in HiC data is to sum biological replicate matrices in order to increase sequencing depth, and thus matrix resolution. This can be done after checking that the biological replicates are indeed similar. It is advised to also conduct downstream analyses separately on each replicate to assess differences at those levels.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"></code></pre></div>
        </div>
      </div>
      </article>
      <div class="footer">
        <a class="label swc-blue-bg" href="mailto:selene.fernandez.valverde@gmail.com">Contacto</a>
        <a class="label swc-blue-bg" href="LICENSE.html">Licencia</a>
      </div>
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
