<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>HiC Workshop: </title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="PBI - Transcriptomica" href="http://bi.ira.cinvestav.mx:8282/"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      <div class="banner">
        <a href="https://liz-fernandez.github.io/HiC-Langebio/" title="HiC Workshop">
          <img alt="Logo Cinvestav-Royal" src="img/cinvestav_Royal.jpg" />
        </a>
      </div>
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <h1 class="title"></h1>
          <hr />
<p>layout: page title: Theoretical and Practical HiC Workshop subtitle: Feature annotation: Compartments, TADs and peaks minutes: 5</p>
<h2 id="compartments">Compartments</h2>
<p>One of the coarsest levels of higher order chromatin organization is the spatial segregation of transcriptionally active and inactive genomic regions. In the earliest HiC maps, these regions were identified as a checker board pattern in the off diagonal. Two global compartments were proposed: A compartment of increased interactions among active regions, and B compartment of inactive regions.</p>
<p>The checker board pattern implies that active regions have a similar interaction profile, so their correlation should be higher than with inactive regions.</p>
<p>We can inspect the pearson correlation matrix to visualize this concept.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicTransform</span> --method pearson -m ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool --outFileName pearson.cool

<span class="kw">hicPlotMatrix</span> -m pearson.cool -o plot_pearson.png --region 2:120000000-132000000 </code></pre></div>
<p>Principal Component Analysis is a dimentionality reduction technique that finds linear combinations of the original matrix variables that a) maximize the variance, b) are independent, and c) are ordered by the percentage of explained variance. This means that applying PCA can be useful to reduce the global correlation patterns to a single vector that captures most of the variability.</p>
<p>The A / B compartments were originally defined in terms of the first component of the matrix. For example, regions with a positive value were assigned to the A compartment, while regions with a negative one to the B compartment.</p>
<p>We can use the hicPCA command to obtain this principal component.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicPCA</span> -noe 1 –matrix ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool  --format bigwig -o pca1.bw</code></pre></div>
<p>Now let’s plot the first principal component beside the matrix.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicPlotMatrix</span> -m ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool -o plot_compartments.png --log1p --region 2:120000000-132000000 --bigwig pca1.bedgraph</code></pre></div>
<p>To emphasize compartmentalization, we can also plot the pearson correlation matrix.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicTransform</span> --method pearson -m ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool --outFileName pearson.cool


<span class="kw">hicPlotMatrix</span> -m pearson.cool -o plot_pearson.png --region 2:120000000-132000000 --bigwig pca1.bw</code></pre></div>
<p>The first component sign only tells us which regions are interacting more frequently with which others, so we can assign two compartments this way. However, from the interaction frequency alone we cannot tell wether a region is transcriptionally active or not.</p>
<p>For this reason, additional information is often used to assign the A and B status to the compartments. RNA seq and histone modification ChIP tracks can be used for this purpose.</p>
<p>Let’s make a new plot to add histone modification information (H3K27me3, which should mark the B compartment).</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicPlotMatrix</span> -m pearson.cool -o plot_pearson.png --region 2:120000000-132000000 --bigwig pca1.bw --bigwig ZmMC_H3K27me3.bw</code></pre></div>
<h2 id="tads">TADs</h2>
<p>Topologically Associating Domains are defined as regions of increased self interaction in HiC maps. Several molecular mechanisms have been proposed for how they can arise, and many computational tools have been developed to identify them.</p>
<p>To call TADs using HiCExplorer we can use the hicFindTADs tool. For each bin along the diagonal, hicFindTADs computes a TAD separation score. This score can be interpreted as the amount of contacts that each bin has with other upstream and downstream bins.</p>
<p>In particular, hicFindTADs works in the following way: - The corrected matrix is transformed to a Z score matrix, based on the average contact frequency at the same genomic distance. - The average Z score within a window of upstream and downstream bins of length w is calculated for each bin. This results in a “diamond” submatrix. - The previous step is repeated for multiple values of w, and an average per bin is computed. This is the TAD separation score. - The local minima of the resulting score track are putative TAD boundaries. Note that the TAD separation score can be thought of as the “boundary strength”. - To double check that these are likely boundaries, for each putative boundary bin, the distribution of Z scores within its diamond submatrix, is compared to the upstream and downstream diamond submatrices.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicFindTADs</span> -m ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool --outPrefix hic_tads --numberOfProcessors 4</code></pre></div>
<h2 id="peaks">Peaks</h2>
<p>To detect peaks, we input the corrected matrix.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicDetectLoops</span> -m ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool -o loops.bedgraph --maxLoopDistance 20000000 --windowSize 10 --peakWidth 6 --pValuePreselection 0.05 --pValue 0.05 --peakInteractionsThreshold 20 --maximumInteractionPercentageThreshold 0.1 --statisticalTest anderson-darling</code></pre></div>
<p>We can plot the matrix with annotated loops.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicPlotMatrix</span> -m ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool -o plot.png --log1p --region 1:18000000-22000000 --loops loops.bedgraph</code></pre></div>
<p>To get an idea of the quality of the callset, besides plotting the annotated matrix we can also produce a plot showing the average of all the called peaks.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hicAggregateContacts</span> --matrix  ZmMC_HiC_2.1.10_sub_1_2.hicup.1mb.corrected.cool --BED loops.bedgraph --outFileName ZmMC_HiC_peaks --vMin 0.8 --vMax 2.2 --range 300000:1000000 --numberOfBins 30 --chromosomes 2 --avgType mean --transform obs/exp</code></pre></div>
<p>In the maize Dong et al. paper, they call peaks using HiCCUPs.</p>
<p>To use this juicer tool we need the .hic matrix.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">hiccups</span> --cpu --threads 4 -c 2 -r 10000 ZmMC_HiC_2.1.10_sub_1_2.hicup.hic -k KR maize_hiccups_loops</code></pre></div>
<p>Now let’s look at the aggregated peak plot.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">apa</span> -r 10000 -c 2 ZmMC_HiC_2.1.10_sub_1_2.hicup.hic maize_hiccups_loops maize_hiccups_apa </code></pre></div>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge"><span class="glyphicon glyphicon-pencil"></span>Challenge:</h2>
</div>
<div class="panel-body">
<ul>
<li>Visualize the resulting peaks in juicebox.</li>
</ul>
</div>
</section>
        </div>
      </div>
      </article>
      <div class="footer">
        <a class="label swc-blue-bg" href="mailto:selene.fernandez.valverde@gmail.com">Contacto</a>
        <a class="label swc-blue-bg" href="LICENSE.html">Licencia</a>
      </div>
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
